[MI-2501] Added unit tests for Carousel input and moved constants to separate package

manojmalik20 opened the pull request:
Added unit tests for the HandlePickerSelection API for the carousel section
Added unit tests for new functions in kv.go
Added unit tests for functions in virtual_agent.go file related to carousel input

manojmalik20 committed:
[MI-2501] Added unit tests for Carousel input and moved constants to separate package
Added unit tests for the HandlePickerSelection API for the carousel action
Added unit tests for new functions in kv.go
Added unit tests for functions in virtual_agent.go file related to carousel input

ayusht2810 commented on server/constants/routes.go:
```diff
@@ -0,0 +1,16 @@
+	PathSetDateTimeDialog          = "/date_time"
+	PathSetDateTime                = "/selected_date_time"
+
+	PathParamEncryptedFileInfo = "encryptedFileInfo"
```
I think the path param should be in constants.go file

ayusht2810 commented on server/plugin/kv_test.go:
```diff
@@ -80,3 +81,34 @@ func Test_StoreUser(t *testing.T) {
+			if testCase.loadError == nil {
+				require.Nil(t, err)
+			} else {
+				require.NotNil(t, err)
```
we can check for error value here as well

ayusht2810 commented on server/plugin/virtual_agent_test.go:
```diff
@@ -452,3 +455,120 @@ func Test_CreateMessageAttachment(t *testing.T) {
+		setupStore    func(s *mock_plugin.MockStore)
+	}{
+		{
+			description: "carousel is handled successfully and there is no error",
```
```suggestion
			description: "carousel is handled successfully with no error",
```

ayusht2810 commented on server/plugin/virtual_agent_test.go:
```diff
@@ -452,3 +455,120 @@ func Test_CreateMessageAttachment(t *testing.T) {
+			test.setupStore(mockedStore)
+			p.store = mockedStore
+
+			p.handlePreviousCarouselPosts(testutils.GetID())
```
don't we need to check something here?

manojmalik20 replied:
No, we don't. If you see the function definition, there's nothing to check. Besides, all the assertions are already happening when we are calling `Expect` function of the mocks.

manojmalik20 committed:
[MI-2501] Modified some tests and fixed lint errors

ayusht2810 commented on server/plugin/kv_test.go:
```diff
@@ -108,6 +109,7 @@ func Test_LoadPostIDs(t *testing.T) {
 				require.Nil(t, err)
 			} else {
 				require.NotNil(t, err)
```
we can remove this line now as we are checking the value now

manojmalik20 committed:
Merge branch 'epic_phase_3' of github.com:Brightscout/mattermost-plugin-servicenow-virtual-agent into MI-2501

manojmalik20 committed:
[MI-2501] Moved a function to utils and wrote unit tests for that

manojmalik20 committed:
[MI-2501] Minor review fix

Kshitij-Katiyar commented on server/plugin/api_test.go:
```diff
@@ -651,6 +676,65 @@ func TestPlugin_handlePickerSelection(t *testing.T) {
 			ParseAuthTokenErr: nil,
 			LoadUserErr:       nil,
 		},
+		"Picker is carousel type and error occurs while getting post": {
```
```suggestion
		"Picker is carousel type and the error occurs while getting post": {
```

Kshitij-Katiyar commented on server/plugin/kv_test.go:
```diff
@@ -80,3 +82,35 @@ func Test_StoreUser(t *testing.T) {
+			description: "Post IDs are loaded successfully from the KV store",
+		},
+		{
+			description: "Post IDs are not loaded successfully from the KV store",
```
```suggestion
			description: "Post IDs are not loaded from the KV store",
```

avas27JTG commented on server/plugin/utils.go:
```diff
@@ -38,8 +44,17 @@ func (p *Plugin) validateTime(time string) string {
+func (p *Plugin) IsCharCountSafe(attachments []*model.SlackAttachment) bool {
+	bytes, err := json.Marshal(attachments)
+	if err != nil {
+		p.API.LogDebug("Error in marshaling the attachments", "Error", err.Error())
```
should not we return here in case of error?

manojmalik20 replied:
I don't even think that there's a need to handle the error here, so reverting this.

avas27JTG commented on server/plugin/utils_test.go:
```diff
@@ -114,3 +117,32 @@ func Test_validateTime(t *testing.T) {
+			description: "result is returned successfully with no error",
+		},
+		{
+			description:  "result is returned successfully with an error",
```
why is it designed this way that both `result` and `error` are returned?
or may be we can make it more clear with `result is returned as false with an error`

avas27JTG commented on server/plugin/virtual_agent_test.go:
```diff
@@ -452,3 +455,120 @@ func Test_CreateMessageAttachment(t *testing.T) {
+			},
+		},
+		{
+			description: "error in storing postIDs and getting post from API",
```
```suggestion
			description: "error occurred while storing postIDs and getting post from API",
```
I think `while` sounds more clear than `in`, let me know your opinion

manojmalik20 committed:
Merge branch 'epic_phase_3' of github.com:Brightscout/mattermost-plugin-servicenow-virtual-agent into MI-2501

manojmalik20 committed:
[MI-2501] Removed return statements after the calls to ScheduleJob and fixed failing unit tests
Modified the util IsCharCountSafe as there was no need to check for errors
Modified some unit tests according to the ScheduleJob change
